llvm/lib/CodeGen/MachineVerifier.cpp

```diff
@@ -1274,6 +1274,12 @@ void MachineVerifier::verifyPreISelGenericInstruction(const MachineInstr *MI) {
         if (TypeSize::isKnownGT(MMO.getSize().getValue(),
                                 ValTy.getSizeInBytes()))
           report("load memory size cannot exceed result size", MI);
+
+        if (MMO.getRanges() && (ValTy.isVector() != MMO.getType().isVector())) {
```

Comment:
This isn't quite right. You're checking a different condition than #97290 is asking for. Neither of the checked types are from MMO.getRanges. We probably should have a similar check, without requiring ranges, but I believe we're relying on the current behavior so it's beyond the scope of this patch.

Instead you should be verifying the IR type in MMO.getRanges() against ValTy and MMO.getType()

Reply:
This is something I likely do not understand,
I can get a type from range for example like this:
`mdconst::dyn_extract<ConstantInt>(MMO.getRanges()->getOperand(0))->getType()`
but there is no other option but ConstantInt, so it cannot be vector, and that was the reason I checked for range presence because I was thinking it indicates scalar nature of the value.
Let's assume, I have this type now, what should I check for, should int type of the range be equal to ValTy?
Should I check all three to be the same? Or should I just check for vector vs scalar mismatches?

Thank you for your help!

Reply:
You get the scalar integer type out of `mdconst::extract<ConstantInt>(MMO.getRanges()->getOperand(0))->getType()`. You need to verify the MIR result type has the same scalar type, something like `ValTy.getScalarType()->getSizeInBits() == mdconst::extract<ConstantInt>(MMO.getRanges()->getOperand(0)).getIntegerType()->getBitWidth()`. Whether the MIR type is vector isn't as significant as whether the element type matches the range IR type

Reply:
I am not sure I understood it correctly, if I get the bitWidth, it will be 64 in all three cases and types will be compatible, but if I take actual bits that are used by lower and higher bounds of the range, it breaks existing tests. Please take a look at the updated code and let me know what am I missing. Thank you for your patience!

Reply:
This set of IR tests happens to only use 64-bit values 

> , it breaks existing tests.

It's possible there are broken tests or bugs, that's why we write these verifiers